orebfuscator-plugin/src/main/java/net/imprex/orebfuscator/injector/OrebfuscatorInjector.java

```diff
@@ -0,0 +1,168 @@
+import net.imprex.orebfuscator.obfuscation.ObfuscationSystem;
+import net.imprex.orebfuscator.player.OrebfuscatorPlayer;
+import net.imprex.orebfuscator.player.OrebfuscatorPlayerMap;
+import net.imprex.orebfuscator.util.BlockPos;
+import net.imprex.orebfuscator.util.OFCLogger;
+import net.imprex.orebfuscator.util.PermissionUtil;
+
+public class OrebfuscatorInjector {
+
+	private static final String OUTBOUND_HANDLER_NAME = "orebfuscator_outbound";
+
+	private final OrebfuscatorConfig config;
+	private final OrebfuscatorPlayerMap playerMap;
+	private final ObfuscationSystem obfuscationSystem;
+
+	private final Player player;
+	private final Channel channel;
+
+	private final Queue<Long> delayPackets = new LinkedList<Long>();
+
+	public OrebfuscatorInjector(Orebfuscator orebfuscator, Player player) {
+		this.config = orebfuscator.getOrebfuscatorConfig();
+		this.playerMap = orebfuscator.getPlayerMap();
+		this.obfuscationSystem = orebfuscator.getObfuscationSystem();
+
+		Object networkManager = MinecraftFields.getNetworkManager(player);
+		var channelAccessor = Accessors.getFieldAccessor(FuzzyReflection.fromObject(networkManager, true)
+				.getField(FuzzyFieldContract.newBuilder()
+				.typeExact(Channel.class)
+				.banModifier(Modifier.STATIC)
+				.build()));
+
+		this.player = player;
+		this.channel = (Channel) channelAccessor.get(networkManager);
+
+		ChannelPipeline pipeline = this.channel.pipeline();
+
+		String encoderName = pipeline.get("outbound_config") != null
+				? "outbound_config"
+				: "encoder";
+
+		if (pipeline.context(OUTBOUND_HANDLER_NAME) == null) {
+			pipeline.addAfter(encoderName, OUTBOUND_HANDLER_NAME, new AsyncOutboundPacketHandler(this));
+		}
+	}
+
+	public void uninject() {
+		if (!this.channel.eventLoop().inEventLoop()) {
+			channel.eventLoop().execute(this::uninject);
+			return;
+		}
+
+		ChannelPipeline pipeline = this.channel.pipeline();
+		if (pipeline.context(OUTBOUND_HANDLER_NAME) != null) {
+			pipeline.remove(OUTBOUND_HANDLER_NAME);
+		}
+	}
+
+	public void logPacketDelay(long nanos) {
+		while (this.delayPackets.size() > 10_000)
+			this.delayPackets.poll();
+		
+		this.delayPackets.offer(nanos);
+	}
+
+	public double averagePacketDelay() {
+		return this.delayPackets.stream()
+				.mapToLong(Long::longValue)
+				.average()
+				.orElse(0d);
+	}
+
+	public PacketType.Protocol getOutboundProtocol() {
+		return ChannelProtocolUtil.PROTOCOL_RESOLVER.apply(this.channel, PacketType.Sender.SERVER);
+	}
+
+	public boolean hasOutboundAsyncListener(PacketType packetType) {
+		return packetType == PacketType.Play.Server.MAP_CHUNK;
+	}
+
+	public void processOutboundAsync(PacketType packetType, Object packet, Promise<Object> promise) {
+		if (this.shouldNotObfuscate(this.player)) {
+			return;
+		}
+
+		PacketContainer packetContainer = new PacketContainer(packetType, packet);
+		ChunkStruct struct = new ChunkStruct(packetContainer, this.player.getWorld());
```

Comment:
I dont like that we still get world in async. Maybe you can use sync ProtocolLib listener just to map `System.identityHashCode(packet.getHandle())` -> `player.getWorld()`  and here try to get it from map, if null fallback to player.getWorld().

Also i think sync listener could be used for starting obfuscate proccess for chunk as soon as possible, so we have minimal packet delay(Mostly is starting getting neighboring chunks as soon as posible (i think it is main source for packet delay)). At least on modern versions if you try to get chunk async while being on main thread it firstly check if chunk is already loaded and if loaded then it completes future immediately so this behaviour can be used to reduce packet delay.

Reply:
Honestly a good idea would like to implement this maybe at a later stage not sure yet, I have to think about a possible implementation a bit.